### PR TITLE
[JUJU-294] Various backup fixes

### PR DESCRIPTION
**Pull request by wallyworld**
This PR fixes several issues with backups:
- properly handle non-default backup dir for juju-db snap
- fixes for downloading backups
- remove option to store backups in mongo
- remove juju CLI and server side code which operated on backups in mongo

There's also a bunch of test cleanup (a little more needs doing but it's a good start). The api tests have JujuConnSuite removed.
Some of the changes are also due to moving the code to check for disk space new to the code which calculates the backup size. It much tidier now. 

When using a non-default backup-dir and juju-db from snap, the processing of the staged files was not quite right.
For at least 3 years, we have defaulted to *not* storing the juju backup archive in mongo. Doing so opens a world of pain. To avoid that footgun, the option to do so is removed. This also means that juju commands to list, remove, show, and restore juju backups in mongo can also be removed.

We are left with 2 commands:
- create-backup
- download-backup

Both will operate on the backup archive which has been stored on disk. Creating a backup will give the full remote filename which can be used to download later (if `--no-download` was specified).

For older juju clients, api stubs are left behind to print a message advising that backups on the controller are not supported.

## QA steps

Using a juju 2.8 client
```
juju list-backups
ERROR backups stored in the controller database not supported (not supported)
```
Using this PR (test with a controller on bionic and also focal)
example below is for juju-db snap
```
juju model-config backup-dir=/root/snap/juju-db/common
juju create-backup --no-download
WARNING downloading backup archives is recommended; backups stored remotely are not guaranteed to be available.

backup format version: 1 
juju version:          2.9.22.2 
series:                focal 

controller UUID:       129f2b25-7496-4053-8b5f-b76d9ec04ba9
model UUID:            3d6c1022-c1fc-443c-865a-1aec27e5c20d 
machine ID:            0 
created on host:       juju-e5c20d-0 

checksum:              ceVUZqG3uT+FPaiDIMieiu9jpDA= 
checksum format:       SHA-1, base64 encoded 
size (B):              233115513 
stored:                0001-01-01 00:00:00 +0000 UTC 
started:               2021-12-03 01:31:04.59200793 +0000 UTC 
finished:              2021-12-03 01:31:36.694963504 +0000 UTC 

notes:                  

Remote backup stored on the controller as /root/snap/juju-db/common/juju-backup-20211203-013104.tar.gz

juju download-backup /root/snap/juju-db/common/juju-backup-20211203-013104.tar.gz
juju-backup-20211203-013104.tar.gz
```
The download above removes the backup from the controller.
Testing create with implicit download
```
juju create-backup

backup format version: 1 
juju version:          2.9.22.5 
series:                focal 

controller UUID:       129f2b25-7496-4053-8b5f-b76d9ec04ba9
model UUID:            3d6c1022-c1fc-443c-865a-1aec27e5c20d 
machine ID:            0 
created on host:       juju-e5c20d-0 

checksum:              PHY+w/h9TMEGpZkvdHNGG0DORN8= 
checksum format:       SHA-1, base64 encoded 
size (B):              688338211 
stored:                0001-01-01 00:00:00 +0000 UTC 
started:               2021-12-03 02:02:48.62034801 +0000 UTC 
finished:              2021-12-03 02:04:12.851832017 +0000 UTC 

notes:                  

Downloaded to juju-backup-20211203-020248.tar.gz
``` 

## Bug reference

https://bugs.launchpad.net/bugs/1950546

### CODE REVIEW

File: apiserver/backup.go [hunk 1]
```diff
@@ -12,20 +12,12 @@ import (
...
 
-var newBackups = func(st *state.State, m *state.Model) (backups.Backups, io.Closer) {
-	backend := struct {
-		*state.State
-		*state.Model
-	}{st, m}
-	stor := backups.NewStorage(backend)
-	return backups.NewBackups(stor), stor
+var newBackups = func() backups.Backups {
+	return backups.NewBackups()
```
Comment (tlm) [inline apiserver/backup.go:20]: what about `newBackups := backups.Newbackups`

File: apiserver/facades/client/backups/backups.go [hunk 2]
```diff
@@ -131,9 +126,8 @@ func extractResourceValue(resources facade.Resources, key string) (string, error
 	return strRes.String(), nil
 }
 
-var newBackups = func(backend Backend) (backups.Backups, io.Closer) {
-	stor := backups.NewStorage(backend)
-	return backups.NewBackups(stor), stor
+var newBackups = func(backend Backend) backups.Backups {
+	return backups.NewBackups()
```
Comment (tlm) [inline apiserver/facades/client/backups/backups.go:130]: Same again

File: state/backups/backups.go [hunk 3]
```diff
@@ -206,67 +200,22 @@ func (b *backups) Create(meta *Metadata, paths *Paths, dbInfo *DBInfo, keepCopy,
...
-		if err2 := build.removeRootDir(); err2 != nil {
-			logger.Errorf(err2.Error())
-			if err != nil {
-				err = errors.Annotatef(err2, "getArchiveFromFilename(%s)", name)
-			}
+		// On success, remove the retrieved file.
+		if err != nil {
+			return
+		}
+		if err2 := os.Remove(fileName); err2 != nil && !os.IsNotExist(err2) {
```
Comment (tlm) [inline state/backups/backups.go:213]: why err2 here? scope changes.

**Review by tlm — APPROVED** [review]

File: state/backups/backups.go (hunk 3 above)
Comment (wallyworld) [reply state/backups/backups.go:213]: yeah, the original code needed a separate err var but the refactored code could do away with it

File: apiserver/backup.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/backup.go:20]: yeah good point, the func was needed before but not now

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**